### DISCUSSION TITLE
Support Enum as Route parameter

### DISCUSSION THREAD

**Discussion by aurimasniekis** [symfony/symfony#44251] [Q&A]
With the new PHP 8.1 releasing soon, I found that in many spots where currently route parameter is some kind of type, format, action (string/int) value could really benefit from upcoming Enum support.

```php
enum Status: string
{
    case DRAFT = 'draft';
    case PUBLISHED = 'published';
    case ARCHIVED = 'archived';
}

#[Route("/route/{id}/{status}")]
public function action(string $id, Status $status): Response {
    // ...
}
```

I am just not sure how to implement this kind of feature as action arguments are parsed after the route is matched and we would need to somehow inject route requirements into the route before routes are compiled into something like this:

```php
#[Route("/route/{id}/{status<^(draft|published|archived)$>}")]
```

It would definitely be really beneficial to support Enums in route parameters would make making strictly typed applications even easier.

**Comment by stof** [comment]
I think this should not be implemented as part of the routing, but as part of the ArgumentResolver system here (the routing won't know the enum type at all to be able to convert it during matching).

For route generation, it might make sense to add support for BackedEnum to convert them automatically to their value (but the workaround for that part is to make your enum stringable as the UrlGenerator already supports stringable objects)

**Reply by stof** [reply]
Note that the argument resolver part should be about implementing a `BackedEnumArgumentResolver` supporting backed enums. For non-backed enums, you will need a custom resolver

**Reply by aurimasniekis** [reply]
Yeah, you can already make something similar even before Enum was released with ArgumentResolver, but my suggestion is somehow to define that parameter as Enum on the route. I was thinking maybe specify as requirement value a class name or something.

**Reply by aurimasniekis** [reply]
The key point is a requirement of parameter not the conversion of value to Enum object.

**Reply by stof** [reply]
ah, if this is about the route requirement, this might indeed be a case where integrating with string backed enums make sense. But this will need some thinking to see how this can work (not only when using attributes)

**Reply by aurimasniekis** [reply]
Yes, YAML/XML configuration files would require some kind of solution which I can't figure out without just using the full class name with a namespace or something.
